Use Vesk for authentication randomness on musl.
Remove libbsd autodetection and all weak fallback paths. A void key API
must terminate rather than return uninitialized or predictable output.

diff --git a/Key.c b/Key.c
index 3a6d669..47f9751 100644
--- a/Key.c
+++ b/Key.c
@@ -32,68 +32,15 @@ in this Software without prior written authorization from The Open Group.
 #include <X11/Xmd.h>
 #include <X11/Xdmcp.h>
 #include <stdlib.h>
-
-#ifndef HAVE_ARC4RANDOM_BUF
-static void
-getbits (long data, unsigned char *dst)
-{
-    dst[0] = (data      ) & 0xff;
-    dst[1] = (data >>  8) & 0xff;
-    dst[2] = (data >> 16) & 0xff;
-    dst[3] = (data >> 24) & 0xff;
-}
-
-#define Time_t time_t
-
-#if defined(HAVE_LRAND48) && defined(HAVE_SRAND48)
-#define srandom srand48
-#define random lrand48
-#endif
-#ifdef WIN32
-#include <process.h>
-#define srandom srand
-#define random rand
-#define getpid(x) _getpid(x)
-#endif
-
-/* Solaris 11.3.0 - 11.4.15 only define getentropy() in <sys/random.h> */
-#if HAVE_GETENTROPY && HAVE_SYS_RANDOM_H
-# include <sys/random.h>
-#endif
-
-static void
-insecure_getrandom_buf (unsigned char *auth, int len)
-{
-    long    lowbits, highbits;
-
-    srandom ((int)getpid() ^ time((Time_t *)0));
-    lowbits = random ();
-    highbits = random ();
-    getbits (lowbits, auth);
-    getbits (highbits, auth + 4);
-}
-
-static void
-arc4random_buf (void *auth, int len)
-{
-#if HAVE_GETENTROPY
-    int	    ret;
-
-    /* weak emulation of arc4random through the getentropy libc call */
-    ret = getentropy (auth, len);
-    if (ret == 0)
-	return;
-#endif /* HAVE_GETENTROPY */
-
-    insecure_getrandom_buf (auth, len);
-}
-
-#endif /* !defined(HAVE_ARC4RANDOM_BUF) */
+#include <unistd.h>
+#include <openssl/rand.h>
 
 void
 XdmcpGenerateKey (XdmAuthKeyPtr key)
 {
-    arc4random_buf(key->data, 8);
+    /* This void API cannot report failure. Never return an unusable key. */
+    if (RAND_bytes(key->data, sizeof(key->data)) != 1)
+	_exit(1);
 }
 
 int
diff --git a/Makefile.am b/Makefile.am
index 3c45924..cf3e382 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -4,11 +4,12 @@ lib_LTLIBRARIES = libXdmcp.la
 
 AM_CPPFLAGS = -I${top_srcdir}/include
 AM_CFLAGS = \
+	$(VESK_CFLAGS) \
 	$(XDMCP_CFLAGS) \
 	$(CWARNFLAGS)
 
 libXdmcp_la_LDFLAGS = -version-number 6:0:0 -no-undefined
-libXdmcp_la_LIBADD = $(XDMCP_LIBS)
+libXdmcp_la_LIBADD = $(XDMCP_LIBS) $(VESK_LIBS)
 libXdmcp_la_SOURCES = \
 	Array.c \
 	Fill.c \
diff --git a/configure.ac b/configure.ac
index 61df441..350fb8d 100644
--- a/configure.ac
+++ b/configure.ac
@@ -53,9 +53,6 @@ XORG_CHECK_SGML_DOCTOOLS(1.8)
 # Checks for programs.
 AC_PROG_LN_S
 
-# Checks for header files.
-AC_CHECK_HEADERS([sys/random.h])
-
 # Checks for libraries.
 AC_SEARCH_LIBS([recvfrom],[socket])
 
@@ -67,19 +64,8 @@ case $host_os in
         ;;
 esac
 
-# Checks for non-standard functions and fallback to libbsd if we can
-# We only check for arc4random_buf, because if we have that, we don't
-# need/use getentropy.
-AC_LINK_IFELSE([AC_LANG_CALL([], [arc4random_buf])],
-               [TRY_LIBBSD="no"], [TRY_LIBBSD="yes"])
-AS_IF([test "x$TRY_LIBBSD" = "xyes"],
-      [PKG_CHECK_MODULES([LIBBSD], [libbsd-overlay], [
-	CFLAGS="$CFLAGS $LIBBSD_CFLAGS"
-	LIBS="$LIBS $LIBBSD_LIBS"
-], [:])])
-
-# Checks for library functions.
-AC_CHECK_FUNCS([srand48 lrand48 arc4random_buf getentropy])
+# Authentication randomness has one provider, with no weak fallback.
+PKG_CHECK_MODULES([VESK], [vesk-crypto])
 
 # Obtain compiler/linker options for dependencies
 PKG_CHECK_MODULES(XDMCP, xproto)
diff --git a/test/KeyRandom.c b/test/KeyRandom.c
new file mode 100644
index 0000000..c92b115
--- /dev/null
+++ b/test/KeyRandom.c
@@ -0,0 +1,42 @@
+#include <X11/Xdmcp.h>
+#include <openssl/rand.h>
+#include <string.h>
+#include <sys/wait.h>
+#include <unistd.h>
+
+static int rng_result = 1;
+
+int
+RAND_bytes(unsigned char *buf, int len)
+{
+    memset(buf, 0xa5, len);
+    return rng_result;
+}
+
+int
+main(void)
+{
+    XdmAuthKeyRec key;
+    int status;
+    pid_t pid;
+    unsigned int i;
+
+    XdmcpGenerateKey(&key);
+    for (i = 0; i < sizeof(key.data); i++)
+        if (key.data[i] != 0xa5)
+            return 1;
+
+    for (rng_result = 0; rng_result >= -1; rng_result--) {
+        pid = fork();
+        if (pid == -1)
+            return 1;
+        if (pid == 0) {
+            XdmcpGenerateKey(&key);
+            _exit(99);
+        }
+        if (waitpid(pid, &status, 0) != pid ||
+            !WIFEXITED(status) || WEXITSTATUS(status) != 1)
+            return 1;
+    }
+    return 0;
+}
diff --git a/test/Makefile.am b/test/Makefile.am
index d86719e..8b3ac2a 100644
--- a/test/Makefile.am
+++ b/test/Makefile.am
@@ -1,10 +1,10 @@
 if ENABLE_UNIT_TESTS
 
-check_PROGRAMS = Array
+check_PROGRAMS = Array KeyRandom
 
 TESTS=$(check_PROGRAMS)
 
-AM_CFLAGS = $(CWARNFLAGS) $(XDMCP_CFLAGS)
+AM_CFLAGS = $(CWARNFLAGS) $(XDMCP_CFLAGS) $(VESK_CFLAGS)
 AM_CPPFLAGS = -I$(top_srcdir)/include
 LDADD= $(top_builddir)/libXdmcp.la $(XDMCP_LIBS)
 
diff --git a/xdmcp.pc.in b/xdmcp.pc.in
index 095d172..2fcadb1 100644
--- a/xdmcp.pc.in
+++ b/xdmcp.pc.in
@@ -7,5 +7,6 @@ Name: Xdmcp
 Description: X Display Manager Control Protocol library
 Version: @PACKAGE_VERSION@
 Requires: xproto
+Requires.private: vesk-crypto
 Cflags: -I${includedir}
 Libs: -L${libdir} -lXdmcp
